### PR TITLE
Use pycosat for satisfiability

### LINKED ISSUE #9728: Add an optional dependency of pycosat
We should use [this](https://github.com/ContinuumIO/pycosat) Python binding to picosat solver.

Thoughts?

### PR DESCRIPTION
**Pull request by debugger22**
- [x] Run tests with and without pycosat
- [x] Benchmarks
- [ ] Update release notes

Closes https://github.com/sympy/sympy/issues/9728

### CODE REVIEW

**Comment by debugger22** [comment]
I ran asv on this branch. Here's the result:

```
· Running 18 total benchmarks (1 commits * 1 environments * 18 benchmarks)
[  0.00%] · For sympy commit hash 3430c271:
[  0.00%] ·· Building for py2.7-fastcache-mpmath....
[  0.00%] ·· Benchmarking py2.7-fastcache-mpmath
[  5.56%] ··· Running dsolve.TimeDsolve01.time_dsolve                                                       880.73ms
[ 11.11%] ··· Running integrate.TimeIntegration01.time_doit                                                 239.93ms
[ 16.67%] ··· Running integrate.TimeIntegration01.time_doit_meijerg                                          73.80ms
[ 22.22%] ··· Running large_exprs.TimeLargeExpressionOperations.peakmem_jacobian_wrt_functions                   34M
[ 27.78%] ··· Running large_exprs.TimeLargeExpressionOperations.peakmem_jacobian_wrt_symbols                     34M
[ 33.33%] ··· Running large_exprs.TimeLargeExpressionOperations.peakmem_subs                                     34M
[ 38.89%] ··· Running large_exprs.TimeLargeExpressionOperations.time_cse                                     45.20ms
[ 44.44%] ··· Running large_exprs.TimeLargeExpressionOperations.time_jacobian_wrt_functions                 509.84ms
[ 50.00%] ··· Running large_exprs.TimeLargeExpressionOperations.time_jacobian_wrt_symbols                    10.80ms
[ 55.56%] ··· Running large_exprs.TimeLargeExpressionOperations.time_manual_jacobian_wrt_functions          534.46ms
[ 61.11%] ··· Running large_exprs.TimeLargeExpressionOperations.time_subs                                   309.31ms
[ 66.67%] ··· Running logic.LogicSuite.time_dpll                                                               3.57s
[ 72.22%] ··· Running logic.LogicSuite.time_dpll2                                                           439.72ms
[ 77.78%] ··· Running logic.LogicSuite.time_load_file                                                         7.06ms
[ 83.33%] ··· Running logic.LogicSuite.time_pycosat                                                         449.72ms
[ 88.89%] ··· Running refine.TimeRefine01.time_refine                                                          7.06s
[ 94.44%] ··· Running solve.TimeSolve01.time_solve                                                             4.43s
[100.00%] ··· Running solve.TimeSolve01.time_solve_nocheck                                                     4.41s
```

No difference.

File: .travis.yml [hunk 1]
```diff
@@ -106,6 +106,9 @@ matrix:
     - python: 3.4
       env:
         - TEST_ASCII="true"
+    - python: 3.4
+      env:
+        - TEST_PYCOSAT="true"
```
Comment (asmeurer) [inline .travis.yml]: Is there any way we can avoid adding yet another entry to the Travis test matrix?

**Comment by asmeurer** [comment]
The pycosat tests timed out. Is this slower?

File: .travis.yml (hunk 1 above)
Comment (moorepants) [inline .travis.yml]: Test all optional dependencies together?

**Comment by debugger22** [comment]
This test is slow

```
    from sympy import numbered_symbols
    from sympy.logic.boolalg import Or
    sym = numbered_symbols()
    X = [next(sym) for i in range(100)]
    result = satisfiable(Or(*X), all_models=True)
    for i in range(10):
        assert next(result)
```

File: .travis.yml (hunk 1 above)
Comment (debugger22) [inline .travis.yml]: +1

**Comment by asmeurer** [comment]
The map(list should be unnecessary with the latest version of pycosat.

File: sympy/logic/algorithms/pycosat_wrapper.py [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+            return dict((symbols[abs(lit) - 1], lit > 0) for lit in r)
+        return bool(expr)
+    else:
+        r = pycosat.itersolve(map(list, clauses_int_repr))
+        result = (r != "UNSAT")
+        if not result:
+            return result
+        sols = []
+        # Make solutions sympy compatible
+        for sol in r:
```
Comment (asmeurer) [inline sympy/logic/algorithms/pycosat_wrapper.py]: This is the problem. You shouldn't iterate over all solutions. If necessary, create a new generator for sympy compatibility.

**Comment by debugger22** [comment]
Merged it with matplotlib tests.

**Comment by asmeurer** [comment]
What about the map list thing?

**Comment by debugger22** [comment]
It doesn't work after removing map list. `solve` needs list.

**Comment by asmeurer** [comment]
Is this with the latest version of pycosat? This [should have been fixed](https://github.com/ContinuumIO/pycosat/blob/master/CHANGELOG#L5).

**Comment by asmeurer** [comment]
Ah, I guess it was never uploaded to PyPI. I'll see if that can be fixed.

**Comment by asmeurer** [comment]
@ilanschnell can you upload pycosat 0.6.1 to PyPI?

**Comment by ilanschnell** [comment]
Oh, I guess I never did.  I'll do that later tonight.

**Comment by ilanschnell** [comment]
Pycosat 0.6.1 is now on PyPI

**Comment by debugger22** [comment]
Cool! I'll update the PR.

**Pull request merged**